Add TPC-H query 18 to ParquetTpchTest

9prady9 opened the pull request:
I have modified the existing TPC-H queries 1 and 6 slightly to accommodate adding more queries easily.

I would like add more TPC-H queries, but I first wanted to get the framework changes in place a thumbs up before adding more queries.

9prady9 commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -335,11 +362,163 @@ TEST_F(ParquetTpchTest, q6) {
+        o_orderkey,
+        o_orderdate,
+        o_totalprice,
+        sum(l_quantity) AS quantity
```
I had to add the query as string here because I wasn't sure how to pick the last column while rounding off double values. and whatever I tried didn't work. Any suggestions here are welcome.

9prady9 replied:
I did try this and didn't work.

majetideepak replied:
You cannot use a raw string literal. The following works.
```
  auto duckDbSql = duckDb_->getTpchQuery(18);
  const auto& duckDBRounded = fmt::format(
        "select c_name, c_custkey, o_orderkey, o_orderdate,"
        "round(o_totalprice, 2) as o_totalprice, \"sum(l_quantity)\" from ({})"
        , duckDbSql);
```

9prady9 replied:
This worked 👍🏾 thanks, although I would like to have a string literal in there - I think it makes the SQL statement more readable.

majetideepak commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -335,11 +355,168 @@ TEST_F(ParquetTpchTest, q6) {
+  ASSERT_EQ(5, stats.pipelineStats.size());
+  //// We used the default of 10 splits
+  ASSERT_EQ(30, stats.numFinishedSplits);
+}
```
meta has an internal format check that expects a new line here.

majetideepak commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -36,68 +34,98 @@ namespace fs = std::filesystem;
+
+using TableColumnsMap = std::map<std::string, std::vector<std::string>>;
+
+static const core::SortOrder kAscNullsLast(true, false);
```
We can remove these. Masha recently added these to the core so that we can reuse them.
https://github.com/facebookincubator/velox/blob/main/velox/core/PlanNode.h#L80

majetideepak commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -335,11 +371,109 @@ TEST_F(ParquetTpchTest, q6) {
+  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+  core::PlanNodeId cSrcNodeId, oSrcNodeId, lSrcNodeId;
+
+  auto big_orders = PlanBuilder(planNodeIdGenerator)
```
`bigOrders` (Use camelCase for local variables) https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md

9prady9 replied:
What is preferred way to run this script https://github.com/facebookincubator/velox/blob/main/scripts/run-clang-tidy.py ? Does this warn of such above conventions ?

9prady9 replied:
just regular invocation with file as argument seems to have worked for now, although it gave 68K roughly, don't think they are just about this change.

Any case, I changed the variable name.

majetideepak commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -335,11 +371,109 @@ TEST_F(ParquetTpchTest, q6) {
+
+  auto big_orders = PlanBuilder(planNodeIdGenerator)
+                        .localPartition(
+                            {0},
```
nit: maybe add a comment that `0` here refers to the first column `l_orderkey`

majetideepak commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -335,11 +371,109 @@ TEST_F(ParquetTpchTest, q6) {
+
+  const auto& stats = task->taskStats();
+
+  //// There should be two pipelines
```
nit: `//` here and below

majetideepak reviewed (commented):
Changes look good to me. Just some minor comments. Thanks, @9prady9.

majetideepak commented:
nit: change PR title to: "Add TPC-H query 18 to ParquetTpchTest"?

majetideepak commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -36,68 +34,98 @@ namespace fs = std::filesystem;
+  static std::string writeTableToParquet(
+      const std::string_view& tableName,
+      const std::string_view& parquetWriteSQL,
+      std::shared_ptr<exec::test::TempDirectoryPath> tempDirectory) {
```
In the context of this test, we will always write to `tempDirectory_`. I don't think the third argument is required.

9prady9 replied:
True, I didn't have the tempDirectory as static member earlier. Will remove this.

9prady9 commented:
@mbasmanova I am aware of bigger tables on left hand side of joins, I will look into it.

Also @majetideepak @mbasmanova this is one of the queries among the ones we implemented where the run-times are very slow compared to DuckDB. I would appreciate if you can find any other inefficiencies in the current plan graph.

thank you!

majetideepak commented:
@9prady9 Can you share some details on your setup? TPC-H scale factor, number of drivers in Velox, etc. I assume the file format is Parquet. The current Parquet Reader in Velox has a couple of known performance issues. We are working on quickly resolving them. How does the DWRF format compare?

9prady9 commented:
> 

Here are the details:

- Yes, we used Parquet only till now, haven't used DWRF
- Scale factors: 1, 4, 8 and 10
- Drivers Count = 16 (16 thread CPU)

Is there an active github issue where I can track the progress of this performance issue ?

majetideepak commented:
@9prady9 The issues are not on Github yet. I will file them soon. The DuckDB plan for q18 is below
<img width="1344" alt="Screen Shot 2022-02-21 at 2 45 21 PM" src="https://user-images.githubusercontent.com/5091159/154924916-e3cafaeb-b51b-4add-8626-3386ea0272d2.png">

yingsu00 commented:
> > 
> 
> Here are the details:
> 
> * Yes, we used Parquet only till now, haven't used DWRF
> * Scale factors: 1, 4, 8 and 10
> * Drivers Count = 16 (16 thread CPU)
> 
> Is there an active github issue where I can track the progress of this performance issue ?

@9prady9 Could you share the test results with us? We are very interested in this performance diff.

9prady9 commented:
> > > 
> > 
> > 
> > Here are the details:
> > 
> > * Yes, we used Parquet only till now, haven't used DWRF
> > * Scale factors: 1, 4, 8 and 10
> > * Drivers Count = 16 (16 thread CPU)
> > 
> > Is there an active github issue where I can track the progress of this performance issue ?
> 
> @9prady9 Could you share the test results with us? We are very interested in this performance diff.

@yingsu00 

On my current setup with 16 threads (8 cores) i.e. 16 drivers, for sf = 10, while duckdb ran query 18 in 2.42 seconds, the velox plan ran in 32.60. Hence the request to review the query 18's implementation.

Unfortunately, the current set-up cannot be shared as a standalone example right away. However, you should be able to reproduce the numbers by wrapping the query execution calls in the tests with c++ high res clock. Take an average over few runs to get the better runtime for query 18. You should notice velox's current implementation is slower than duckdb by a good amount. Please try it with sf=10 or higher. The difference may not be that much for lower scale factors I think.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -36,68 +34,93 @@ namespace fs = std::filesystem;
 
+  /// Write the DuckDB TPC-H table to a Parquet file and return the file path
+  ///
+  /// - tableName is name of table from TPC-H tables
```
Use @param to document parameters.

majetideepak replied:
@9prady9 Apologies for asking you to remove `@param` earlier. I went through https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md
and see that it does mention 
```
For multi-line comments:
   Velox will follow the doxygen comment style. 
```

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -36,68 +34,93 @@ namespace fs = std::filesystem;
+  /// - tableName is name of table from TPC-H tables
+  /// - parquetWriteSQL is the string containing SQL query to write the
+  ///   tableName as a parquet file to the disk
+  /// - tempDirectory is the location where parquet files are saved
```
This parameter doesn't exist.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -36,68 +34,93 @@ namespace fs = std::filesystem;
+  /// Returns a std::string containing the path to parquet file on disk
+  static std::string writeTableToParquet(
+      const std::string_view& tableName,
+      const std::string_view& parquetWriteSQL) {
```
nit: createTableSql or createTableQuery name might be clearer

9prady9 replied:
I don't think that matches the action being done inside this static method.  We dumping the table contents of a corresponding TPC-H table into a parquet file. Perhaps `saveTpchTableAsParquetFile` sounds better

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -36,68 +34,93 @@ namespace fs = std::filesystem;
+  ///
+  /// Returns a std::string containing the path to parquet file on disk
+  static std::string writeTableToParquet(
+      const std::string_view& tableName,
```
std::string_view is a small object, so no point in passing it by reference.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -36,68 +34,93 @@ namespace fs = std::filesystem;
+    constexpr int kRowGroupSize = 10'000;
+    const auto& filePath =
+        fmt::format("{}/{}.parquet", tempDirectory_->path, tableName);
+    const auto& query =
```
do not store a reference to a temporary string; use "auto query" instead

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -36,68 +34,93 @@ namespace fs = std::filesystem;
     parquet::unregisterParquetReaderFactory();
   }
 
+  /// Write the DuckDB TPC-H table to a Parquet file and return the file path
```
period at the end of the sentence is missing.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -36,68 +34,93 @@ namespace fs = std::filesystem;
-             VARCHAR()});
+
+    // Convert decimal columns to double.
+    const std::string lineitemSQL = R"(
```
nit: consider extracting helper functions to create different tables, e.g. createLineitem, createOrders and createCustomer, for readability:

```
parquetFilesPaths_["lineitem"] = createLineitem();
parquetFilesPaths_["orders"] = createOrders();
parquetFilesPaths_["customer"] = createCustomer();
```

9prady9 replied:
Unless there is a strong inclination towards such arrangement of code, I honestly don't see the need to pull out these string literals and wrap them with functions unless we plan on using these strings somewhere else.

mbasmanova replied:
This method in its current form is very long and hard to read. As we add more tables it will get longer and longer and harder and harder to read.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,173 @@ class ParquetTpchTest : public testing::Test {
-    duckDb_->execute(query);
-    return filePath;
+  std::shared_ptr<const RowType> getRowType(
+      const std::string& table,
```
consistency: writeTableToParquet takes std::string_view, but this method takes std::string. It would be nice to choose one of these and use them consistently.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,173 @@ class ParquetTpchTest : public testing::Test {
+  std::shared_ptr<const RowType> getRowType(
+      const std::string& table,
+      const std::string& filePath) {
+    auto getReaderRowType = [&](parquet::ParquetReader& reader) {
```
What's the motivation for extracting this logic into a lambda variable? It seems like it can be inlined for better readability.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,173 @@ class ParquetTpchTest : public testing::Test {
+      }
+      return ROW(std::move(names), std::move(types));
+    };
+    dwio::common::ReaderOptions ReaderOptions;
```
naming: ReaderOptions -> readerOptions or options

See coding style: https://github.com/facebookincubator/velox/blob/main/CODING_STYLE.md

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,173 @@ class ParquetTpchTest : public testing::Test {
+      std::vector<std::basic_string<char>> names;
+      std::vector<std::shared_ptr<const Type>> types;
+      for (int i = 0; i < type->size(); i++) {
+        names.push_back(type->type->as<TypeKind::ROW>().nameOf(i));
```
this can be taken out of the loop and shortened to `names = type->type->as<TypeKind::ROW>().names()`

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,173 @@ class ParquetTpchTest : public testing::Test {
+        names.push_back(type->type->as<TypeKind::ROW>().nameOf(i));
+        types.push_back(createScalarType(type->childAt(i)->type->kind()));
+      }
+      return ROW(std::move(names), std::move(types));
```
Is this different from `std::dynamic_pointer_cast<RowType>(type->type)` ?

9prady9 replied:
I didn't dig further, but ROW is a function from this file https://github.com/facebookincubator/velox/blob/main/velox/type/Type.cpp#L504 . Or may be I don't understand your question exactly 🤔

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,173 @@ class ParquetTpchTest : public testing::Test {
+    parquet::ParquetReader reader(
+        std::make_unique<dwio::common::FileInputStream>(filePath),
+        ReaderOptions);
+    return getReaderRowType(reader);
```
Can you use reader->rowType() here?

9prady9 replied:
I couldn't understand, can you please explain further ? `getReaderRowType` was the lambda I had earlier.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,173 @@ class ParquetTpchTest : public testing::Test {
-      types.push_back(lineitemType_->findChild(colName));
-    }
-    return ROW(std::move(names), std::move(types));
+  std::shared_ptr<const facebook::velox::RowType> getRowType(
```
This method would be easier to use if it took a list of columns, not a map from table name to lists of columns, e.g. `tableColumns.find(table)->second` logic seems to belong to the caller.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,173 @@ class ParquetTpchTest : public testing::Test {
       const CursorParameters& params,
-      const std::string& filePath,
-      const core::PlanNodeId& sourcePlanNodeId,
+      const std::vector<std::string> filePaths,
```
pass by reference

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,173 @@ class ParquetTpchTest : public testing::Test {
-      const std::string& filePath,
-      const core::PlanNodeId& sourcePlanNodeId,
+      const std::vector<std::string> filePaths,
+      const std::vector<core::PlanNodeId> sourcePlanNodeIds,
```
ditto

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,173 @@ class ParquetTpchTest : public testing::Test {
-            auto const& splits = makeSplits(filePath);
-            for (const auto& split : splits) {
-              task->addSplit(sourcePlanNodeId, exec::Split(split));
+            for (size_t t = 0; t < filePaths.size(); t++) {
```
naming: t -> i by convention

mbasmanova reviewed (changes requested):
@9prady9 Pradeep, thank you for working on this. Some initial comments.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,169 @@ class ParquetTpchTest : public testing::Test {
+    parquet::ParquetReader reader(
+        std::make_unique<dwio::common::FileInputStream>(filePath),
+        readerOptions);
+    auto type = reader.typeWithId();
```
I wonder if you can simply `return reader.rowType()` instead of re-building the type from `reader.typeWithId()`.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,169 @@ class ParquetTpchTest : public testing::Test {
     return ROW(std::move(names), std::move(types));
   }
 
+  std::shared_ptr<const facebook::velox::RowType> getRowType(
```
nit: use RowTypePtr for readability and drop `facebook::velox::` prefix throughout the diff

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,169 @@ class ParquetTpchTest : public testing::Test {
-       "tax",
-       "shipdate"});
+  const auto& filePath = parquetFilesPaths_["lineitem"];
+  TableColumnsMap tableCols;
```
This map contains only only entry with key "lineitem". You could use and std::vector<std::string> intead.

```
std::vector<std::string> columnNames = {...};
```

do not abbreviate: tableCols -> columnNames

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -123,163 +146,169 @@ class ParquetTpchTest : public testing::Test {
+               DOUBLE(),
+               BIGINT()})
+          .orderBy({0, 1}, {core::kAscNullsLast, core::kAscNullsLast}, false)
+          // Additional step for double type result verification
```
@spershin Sergey is adding logic to allow for slight discrepancies when comparing floating point results in https://github.com/facebookincubator/velox/pull/1108/ . This will allow to remove this workaround. CC: @majetideepak

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -289,39 +318,38 @@ TEST_F(ParquetTpchTest, q1) {
-      getLineitemColumns({"shipdate", "extendedprice", "quantity", "discount"});
-
+  const auto& filePath = parquetFilesPaths_["lineitem"];
+  TableColumnsMap tableCols;
```
same as above

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -335,11 +363,113 @@ TEST_F(ParquetTpchTest, q6) {
+  const auto& ordersFilePath = parquetFilesPaths_["orders"];
+  const auto& customerFilePath = parquetFilesPaths_["customer"];
+
+  TableColumnsMap tableCols;
```
This map is not needed. It would be clearer to use std::vector<std::string> for column names, e.g.

```
std::vector<std::string> lineitemColumns = {...};
std::vector<std::string> ordersColumns = {...};
std::vector<std::string> customerColumns = {...};
```

These can also be inlined:

```
getRowType("lineitem", {"l_orderkey", "l_quantity"}, lineitemFilePath);
...
```

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -335,11 +363,113 @@ TEST_F(ParquetTpchTest, q6) {
+  params.numResultDrivers = 1;
+
+  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+  core::PlanNodeId cSrcNodeId, oSrcNodeId, lSrcNodeId;
```
do not abbreviate; declare each variable on a separate line

```
core::PlanNodeId lineitemScanNodeId;
core::PlanNodeId ordersScanNodeId;
core::PlanNodeId customerScanNodeId;
```

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -335,11 +363,113 @@ TEST_F(ParquetTpchTest, q6) {
+               "round(o_totalprice, cast(2 AS integer))",
+               "quantity"})
+          .planNode();
+  params.planNode = std::move(plan);
```
nit: looks like you don't need plan variable

```
params.planNode = PlanBuilder(planNodeIdGenerator)...
```

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -335,11 +363,113 @@ TEST_F(ParquetTpchTest, q6) {
+
+  const auto& stats = task->taskStats();
+
+  // There should be two pipelines
```
typo: two -> five ?

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -335,11 +363,113 @@ TEST_F(ParquetTpchTest, q6) {
+
+  // There should be two pipelines
+  ASSERT_EQ(5, stats.pipelineStats.size());
+  // We used the default of 10 splits
```
typo: 10 -> 30 ?

mbasmanova reviewed (commented):
@9prady9 Pradeep, thank you for updating the PR. Some follow-up comments below. I haven't reviewed the query plan for q18 yet.

9prady9 commented:
I have addressed the feedback - but I am getting undefined references after rebasing from latest main - I tried `main` branch which also had the problem so, it doesn't seem like the errors are from this PR's changes. Please advice.

mbasmanova commented:
@9prady9 

> I am getting undefined references after rebasing

Would you share the error details? Are these compilation, linking or runtime failures?

mbasmanova commented:
CC: @tanjialiang @gggrace14 @spershin

tanjialiang commented:
> CC: @tanjialiang @gggrace14 @spershin

I'm taking a look

9prady9 commented:
> @9prady9 I am not able to reproduce the errors you got. Could you help to give some repro steps?

I did a clean (make clean and make) build on `main` branch, that is all - no code changes. The ci didn't seem to have a problem, more likely the build folder needs to be totally removed. I can re-try on fresh clone and see if that helps in any way.

tanjialiang commented:
> > @9prady9 I am not able to reproduce the errors you got. Could you help to give some repro steps?
> 
> I did a clean (make clean and make) build on `main` branch, that is all - no code changes. The ci didn't seem to have a problem, more likely the build folder needs to be totally removed. I can re-try on fresh clone and see if that helps in any way.
> 
> Note: I have noticed the failure on one of the jobs, fixing it

Thanks. Please let me know the results of your try. We can continue from there.

9prady9 commented:
> > > @9prady9 I am not able to reproduce the errors you got. Could you help to give some repro steps?
> > 
> > 
> > I did a clean (make clean and make) build on `main` branch, that is all - no code changes. The ci didn't seem to have a problem, more likely the build folder needs to be totally removed. I can re-try on fresh clone and see if that helps in any way.
> > Note: I have noticed the failure on one of the jobs, fixing it
> 
> Thanks. Please let me know the results of your try. We can continue from there.

I can confirm the error is side-effect of corrupt build directory (I think) even though it was after `make clean`. I have updated the plans to remove rounding off operations also.

tanjialiang commented:
> 

Awesome. Thanks for confirming!

mbasmanova commented:
@9prady9 @majetideepak Pradeep, Deepak refactored ParquetTpchTest to extract TpchQueryBuider in https://github.com/facebookincubator/velox/pull/1155 . Would you rebase and update this PR?

9prady9 commented:
@majetideepak What is the idea behind adding aliases ? Isn't it adding lot more code to the builder given that column names in tpch tables isn't going to change

majetideepak commented:
@9prady9 Presto for instance generates TPC-H data with SIMPLIFIED column names. So the aliasing gives more flexibility.

majetideepak commented:
To clarify, simplified column names don't have the prefix `l_` for `lineitem` table columns

mbasmanova commented on velox/exec/tests/utils/TpchQueryBuilder.cpp:
```diff
@@ -268,27 +270,165 @@ TpchPlan TpchQueryBuilder::getQ6Plan() const {
+TpchPlan TpchQueryBuilder::getQ18Plan() const {
+  static const std::vector<std::string> kTableNames = {
+      "lineitem", "orders", "customer"};
+  std::vector<std::string> lineitemSelectedColumns = {
```
nit: consider shortening variable names for readability: lineitemColumns, ordersColumns, customerColumns.

mbasmanova replied:
I think the values clearly show that this is a subset of all columns. Overall, shorter test would allow one to read it better and capture more of these details.

mbasmanova commented on velox/exec/tests/utils/TpchQueryBuilder.cpp:
```diff
@@ -268,27 +270,165 @@ TpchPlan TpchQueryBuilder::getQ6Plan() const {
+  std::vector<std::string> customerSelectedColumns = {"c_name", "c_custkey"};
+
+  // Get the corresponding selected column names in the file.
+  const auto lineitemAliasSelectedColumns =
```
ditto: lineitemAliases, ordersAliases, customerAliases

9prady9 replied:
I agree, these can be shortened because of their localized context I think

mbasmanova commented on velox/exec/tests/utils/TpchQueryBuilder.cpp:
```diff
@@ -268,27 +270,165 @@ TpchPlan TpchQueryBuilder::getQ6Plan() const {
+          tableMetadata_.at(kTableNames[2]).type, customerAliasSelectedColumns);
+
+  auto lineitemSelectedRowType =
+      lineitemColumnSelector->buildSelectedReordered();
```
consider introducing a helper function that takes a list of column names and returns a RowType

```
auto lineitemRowType = getRowType(kLineitem, lineitemColumns)
```

mbasmanova commented on velox/exec/tests/utils/TpchQueryBuilder.cpp:
```diff
@@ -268,27 +270,165 @@ TpchPlan TpchQueryBuilder::getQ6Plan() const {
+                                .capturePlanNodeId(lineitemSourceNodeId)
+                                .partialAggregation({0}, {"sum(l_quantity)"})
+                                .planNode()})
+                       .finalAggregation({0}, {"sum(a0)"}, {DOUBLE()})
```
consider using AS statement in partialAggregation to give a0 a better name

```
.partialAggregation({0}, {"sum(l_quantity) AS sum_quantity"})
...
.finalAggregation({0}, {"sum(sum_quantity)"}, {DOUBLE()})
```

mbasmanova commented on velox/exec/tests/utils/TpchQueryBuilder.cpp:
```diff
@@ -268,27 +270,165 @@ TpchPlan TpchQueryBuilder::getQ6Plan() const {
+                                .partialAggregation({0}, {"sum(l_quantity)"})
+                                .planNode()})
+                       .finalAggregation({0}, {"sum(a0)"}, {DOUBLE()})
+                       .project({"l_orderkey", "a0 AS quantity"})
```
Consider dropping this this node and use AS statement instead

```
.finalAggregation({0}, {"sum(a0) AS quantity"}, {DOUBLE()})
```

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -23,13 +23,23 @@
 #include "velox/functions/prestosql/registration/RegistrationFunctions.h"
 #include "velox/parse/TypeResolver.h"
 
+#include <map>
```
Are these needed?

9prady9 replied:
my mistake, it should be `unordered_map`, some other header must be including this, hence it is not being needed. Should I avoid including it explicitly ?

mbasmanova replied:
> Should I avoid including it explicitly ?

This would be nice.

9prady9 replied:
Okay, removed it but I don't think it is good thing in general to depend on some upstream header to include it.

mbasmanova replied:
Thank you. 

> I don't think it is good thing in general to depend on some upstream header

I agree overall. However, it is impractical to include ALL headers used in a given file as that would require including a lot of headers from standard library for example.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -114,35 +125,50 @@ std::shared_ptr<DuckDbQueryRunner> ParquetTpchTest::duckDb_ = nullptr;
-  ASSERT_EQ(2, stats.pipelineStats.size());
-  // We used the default of 10 splits per file.
-  ASSERT_EQ(10, stats.numFinishedSplits);
+std::ostream& operator<<(std::ostream& os, const QueryParams& params) {
```
consider moving this next to QueryParams structure for readability

mbasmanova replied:
It can be outside, but still needs to be closer to the struct itself, e.g. right after the struct.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -114,35 +125,50 @@ std::shared_ptr<DuckDbQueryRunner> ParquetTpchTest::duckDb_ = nullptr;
-TEST_F(ParquetTpchTest, q6) {
-  auto tpchPlan = tpchBuilder_.getQueryPlan(6);
-  auto duckDbSql = duckDb_->getTpchQuery(6);
+TEST_P(ParquetTpchTest, CompareWithDuckDB) {
```
naming: method names start with lower case; compareWithDuckDB name is not very descriptive as all tests do that. This test verifies correctness of a single TPC-H query, right? Perhaps, "basic" would be a good name?

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -23,13 +23,23 @@
 
-class ParquetTpchTest : public testing::Test {
+struct QueryParams {
+  int queryId;
```
Prefix "query" is redundant as all these variables are scoped to QueryParams struct.

mbasmanova commented on velox/dwio/parquet/tests/ParquetTpchTest.cpp:
```diff
@@ -114,35 +125,50 @@ std::shared_ptr<DuckDbQueryRunner> ParquetTpchTest::duckDb_ = nullptr;
+    CompareWithDuckDB,
+    ParquetTpchTest,
+    testing::Values(
+        QueryParams{1, 2, 10},
```
Using parameterized tests is one way to re-use code. Introducing a helper method would be another way. I don't have a strong preference, but it seems a bit more readable to use a helper method. In particular, it seems easier to pick and choose which tests to run without editing the code. CC: @majetideepak 

```
TEST_P(ParquetTpchTest, q1) {
   doTest(1);
}

TEST_P(ParquetTpchTest, q6) {
   doTest(6);
}

TEST_P(ParquetTpchTest, q18) {
   doTest(18);
}
```

9prady9 replied:
Hm, okay, reverted this change.

mbasmanova reviewed (commented):
@9prady9 Looks good overall. Some comments and questions below.

mbasmanova reviewed (approved):
@9prady9 Pradeep, looks good to me. Thank you for iterating on this PR.
